Misha/cpu scaling

mikhail-sakhnov opened the pull request:
Based on the non-merged PR #1104 



Introduce separate CPU scaling flow based on the vmSpec.cpuScalingMode

If vmSpec.cpuScalingMode is equal to `qmp_scaling` the logic of the
scaling is preserved as before:

- Scale, if required the amount of CPUs using qmp commands.
- If it is required to scale cgroups, call vm-runner /cpu_change endpoint

if vmSpec.cpuScalingMode is equal to `cpuSysfsStateScaling` all cpu
scaling requests go directly to the vm-runner /cpu_change, which in
that configuration goes to the neonvm-daemon to reconcile required
amount of online CPUs.

Value `cpuSysfsStateScaling` also modifies the qemu and the kernel
arguments to enable plug all CPUs but mark as online only first one.

sharnoff commented on neonvm-controller/cmd/main.go:
```diff
@@ -105,6 +106,7 @@ func main() {
 	flag.BoolVar(&enableLeaderElection, "leader-elect", false,
 		"Enable leader election for controller manager. "+
 			"Enabling this will ensure there is only one active controller manager.")
+	flag.BoolVar(&useCpuSysfsStateScaling, "use-cpu-sysfs-state-scaling", false, "Use sysfs cpu state scaling for CPU scaling")
```
nit: this would be better grouped further down, next to the other flags related to settings for the VM

sharnoff commented on Makefile:
```diff
@@ -133,9 +133,10 @@ test: fmt vet envtest ## Run tests.
-	GOOS=linux go build -o bin/controller       neonvm/main.go
-	GOOS=linux go build -o bin/vxlan-controller neonvm/tools/vxlan/controller/main.go
-	GOOS=linux go build -o bin/runner           neonvm/runner/*.go
+	GOOS=linux go build -o bin/controller       neonvm-controller/cmd/main.go
+	GOOS=linux go build -o bin/vxlan-controller neonvm-vxlan-controller/cmd/main.go
+	GOOS=linux go build -o bin/daemon           neonvm-daemon/main.go
+	GOOS=linux go build -o bin/runner           neonvm-runner/cmd/main.go
```
Are these required for this PR? (if no: let's break that out into a separate change? will be easier to discuss without derailing this PR)

mikhail-sakhnov replied:
Not really but I use build target locally for fast builds and run linter

sharnoff replied:
Hm, that sounds like we should potentially improve our local setup there. I always just manually run `go build ./...` or `golangci-lint run` (approximately — the actual commands are much more convoluted, but I have them in my shell history...)

If you also think there's room for improvement, could you open an issue?

sharnoff replied:
We discussed this 1-1 — I'm ok to leave this change in the PR, but maybe we should just remove `make build` entirely (we already have `make lint`, etc.)

sharnoff commented on neonvm-controller/deployment.yaml:
```diff
@@ -55,6 +55,7 @@ spec:
         - "--concurrency-limit=128"
         - "--skip-update-validation-for="
         - "--disable-runner-cgroup"
+        - "--use-cpu-sysfs-state-scaling"        
```
We probably shouldn't enable by default on initial rollout, right? otherwise rollback will result in new computes being unable to be handled by the old controller (?)

sharnoff commented on neonvm-daemon/pkg/cpuscaling/sys_fs_state_scaling.go:
```diff
@@ -0,0 +1,196 @@
+package cpuscaling
```
nits:

* Underscores in filenames have semantic meaning in Go, should we avoid that here?
* The current file structure has a centralized `pkg/` directory but individual `component/cmd/` directories. Can we move this to the top-level `pkg/` directory?

mikhail-sakhnov replied:
Moved, renamed

sharnoff commented on neonvm-daemon/pkg/cpuscaling/sys_fs_state_scaling.go:
```diff
@@ -0,0 +1,196 @@
+// CPU directory path
+const cpuPath = "/sys/devices/system/cpu/"
+
+type CPUSysFsStateScaler struct {
+}
```
What's the intent behind making an empty type with methods, instead of just public functions?

mikhail-sakhnov replied:
Honestly, not a lot of reason to have it struct, but I like less coupled code in the wrapping cpuServer structure much more than just a calls to public functions. 

Plus I was originally going to add unittests here and forget :D

sharnoff commented on neonvm-daemon/pkg/cpuscaling/sys_fs_state_scaling.go:
```diff
@@ -0,0 +1,196 @@
+type CPUSysFsStateScaler struct {
+}
+
+func (c *CPUSysFsStateScaler) EnsureOnlineCPUs(X int) error {
```
```suggestion
func (c *CPUSysFsStateScaler) EnsureOnlineCPUs(x int) error {
```

sharnoff commented on neonvm-daemon/pkg/cpuscaling/sys_fs_state_scaling.go:
```diff
@@ -0,0 +1,196 @@
+		return err
+	}
+
+	if onlineCount < uint32(X) {
+		for _, cpu := range cpus {
+			if cpu == 0 {
```
nit: this is a lot of nested control flow. Maybe worth separating into >1 function?

sharnoff commented on neonvm-daemon/pkg/cpuscaling/sys_fs_state_scaling.go:
```diff
@@ -0,0 +1,196 @@
+}
+
+// parseCPURange parses the CPU range string (e.g., "0-3") and returns a list of CPUs.
+func parseCPURange(cpuRange string) ([]int, error) {
```
Would it make more sense to return `(start ID, end ID, error)` rather than `([]IDs, error)` ?

sharnoff commented on neonvm-daemon/cmd/main.go:
```diff
@@ -26,24 +32,76 @@ func main() {
 
-type cpuServer struct{}
+func (s *cpuServer) handleSetCPUStatus(w http.ResponseWriter, r *http.Request) {
+	// TODO: should the call to this method be conditional, only if the statefs cpu scaling is enabled?
+	// on the other hand, currently this endpoint is called by runner only if the statefs scaling is enabled
+	// and it is a bit tricky to pass vmSpec here
```
commenting on the TODO: it's not worth it to double-check that here. Let's trust what the runner says.

sharnoff commented on neonvm-runner/cmd/main.go:
```diff
@@ -890,12 +889,23 @@ func buildQEMUCmd(
-		vmSpec.Guest.CPUs.Max.RoundedUp(),
-		vmSpec.Guest.CPUs.Max.RoundedUp(),
-	))
+	if vmSpec.CpuScalingMode != nil && *vmSpec.CpuScalingMode == vmv1.CpuScalingModeCpuSysfsState {
```
IMO: we should be moving away from relying on the VM spec in `neonvm-runner`. WDYT about making this a CLI flag instead?

mikhail-sakhnov replied:
Fun fact, that's how it was in the first iteration of the branch 😄 

Sure, will do.

sharnoff commented on neonvm-runner/cmd/main.go:
```diff
@@ -890,12 +889,23 @@ func buildQEMUCmd(
+		// if we use hotplug we specify start cpus equal to min cpus and scale using udev rules for cpu plug events
+		qemuCmd = append(qemuCmd, "-smp", fmt.Sprintf(
+			"cpus=%d,maxcpus=%d,sockets=1,cores=%d,threads=1",
+			vmSpec.Guest.CPUs.Min.RoundedUp(),
+			vmSpec.Guest.CPUs.Max.RoundedUp(),
+			vmSpec.Guest.CPUs.Max.RoundedUp(),
```
nit: this is hard to read, particularly when comparing with the section above it (because "min" vs "max" looks quite similar if the rest of the line is the same :sweat_smile:)

Maybe let's use some helper variables above it?

mikhail-sakhnov replied:
Introduced variables, should be easier to read now

sharnoff commented on neonvm-runner/cmd/main.go:
```diff
@@ -983,6 +993,10 @@ func makeKernelCmdline(cfg *Config, vmSpec *vmv1.VirtualMachineSpec, vmStatus *v
 		cmdlineParts = append(cmdlineParts, cfg.appendKernelCmdline)
 	}
+	if vmSpec.CpuScalingMode != nil && *vmSpec.CpuScalingMode == vmv1.CpuScalingModeCpuSysfsState {
+		// if we use sysfs based scaling we need to specify the start cpus as min CPUs
+		cmdlineParts = append(cmdlineParts, fmt.Sprintf("maxcpus=%d", vmSpec.Guest.CPUs.Min.RoundedUp()))
```
Why is this required in the kernel command line when it wasn't before?

mikhail-sakhnov replied:
Because without that change all cpu would be marked as online on start, while probably we want to have minimum amount of CPUs marked as online on start?

mikhail-sakhnov replied:
so it is what is done for QMP hotplugs by `cpus=minCpus`, has the same semantic

sharnoff replied:
Ahhhh I see. Could you add a comment saying this?

mikhail-sakhnov replied:
it actually has the comment 🤔 I tried to rephrase it to be a more clarifying, https://github.com/neondatabase/autoscaling/pull/1111/files#diff-7eb577493ab5186c29d09f0791bee695a98db31cd3619c95e834a7924744823fR1003

sharnoff commented on neonvm-runner/cmd/main.go:
```diff
@@ -1029,8 +1043,12 @@ func runQEMU(
+	}
 	go terminateQemuOnSigterm(ctx, logger, &wg)
-	if !cfg.skipCgroupManagement || cfg.enableDummyCPUServer {
+	if !cfg.skipCgroupManagement || cfg.enableDummyCPUServer || useCpuSysfsStateScaling {
```
not blocking: @Omrigan with the revert of the crictl patches, this `if` should now be unconditional, right?

sharnoff commented on neonvm/apis/neonvm/v1/virtualmachine_types.go:
```diff
@@ -605,6 +618,7 @@ func (p VmPhase) IsAlive() bool {
 // +kubebuilder:printcolumn:name="Age",type="date",JSONPath=".metadata.creationTimestamp"
 // +kubebuilder:printcolumn:name="Node",type=string,priority=1,JSONPath=`.status.node`
 // +kubebuilder:printcolumn:name="Image",type=string,priority=1,JSONPath=`.spec.guest.rootDisk.image`
+// +kubebuilder:printcolumn:name="CPUScalingMode",type=string,priority=1,JSONPath=`.spec.cpuScalingMode`
```
Is this worth including in `kubectl get neonvm -o wide` ?

sharnoff commented on pkg/api/types.go:
```diff
@@ -476,12 +476,17 @@ type RunnerProtoVersion uint32
 
 const (
 	RunnerProtoV1 RunnerProtoVersion = iota + 1
+	RunnerProtoV2
```
Is this used?

sharnoff reviewed (commented):
Partial review, haven't yet gotten to the neonvm-controller changes

mikhail-sakhnov commented on neonvm-runner/cmd/main.go:
```diff
@@ -1812,3 +1842,33 @@ func overlayNetwork(iface string) (mac.MAC, error) {
 	return mac, nil
 }
+
+func setNeonvmDaemonCPU(cpu vmv1.MilliCPU) error {
```
both options (direct call to vm daemon or through runner) works for me and I don't have strong opinion. @sharnoff I remember you had some points, may I ask you to share them here?

sharnoff replied:
Yeah, broadly my view is that neonvm-daemon is privileged inside the VM, and we shouldn't expose it beyond the pod. With neonvm-runner, it should be _relatively_ easy for us to add mTLS to verify that connections are coming from neonvm-controller, but I'm hesitant to do that in neonvm-daemon because getting secrets into VMs is a little convoluted and has potentially unexpected properties (e.g., they won't be refreshed).

Similarly, in the worst-case scenario of a DoS attack on the HTTP server, I would rather that happens to neonvm-runner than neonvm-daemon so that the internals of the VM are unaffected. There's occasionally attacks on HTTP implementations that require unusual traffic, and having neonvm-runner act as an intermediary guards against that.

mikhail-sakhnov replied:
Actually I just realized that even from perspective of the ownership going through runner is a bit more clear model. Let's say theoretically some operations can be performed only inside vm (throug daemon) but some only outside (through qmp). Why should controller bother to know the details what to use in every particular case?

sharnoff replied:
> Why should controller bother to know the details what to use in every particular case?

Great point, yeah. This is exactly the issue with having the controller directly connect to QEMU for QMP.

mikhail-sakhnov commented on neonvm-daemon/cmd/main.go:
```diff
@@ -26,24 +32,76 @@ func main() {
+	update := vmv1.MilliCPU(updateInt)
+	if err != nil {
+		s.logger.Error("could not unmarshal request body", zap.Error(err))
+		w.WriteHeader(http.StatusBadRequest)
```
It doesn't really change a lot where we log it - here or in the runner, in a both ways log goes to the pod log

mikhail-sakhnov commented on neonvm/apis/neonvm/v1/virtualmachine_types.go:
```diff
@@ -52,6 +52,14 @@ const (
+
+	// CpuScalingModeCpuSysfsState is the value of the VirtualMachineSpec.CpuScalingMode field that
+	// indicates that the VM should use the CPU sysfs state interface to scale CPUs.
+	CpuScalingModeCpuSysfsState string = "sysfsScaling"
```
Renamed the variable, but I don't like changing the value, for the sake of easy grepping

mikhail-sakhnov commented on neonvm-daemon/cmd/main.go:
```diff
@@ -26,24 +32,73 @@ func main() {
+}
+
+func (s *cpuServer) handleGetCPUStatus(w http.ResponseWriter) {
+	// should be replaced with cgroups milliseconds to be exposed instead of having CPU
```
leftover

mikhail-sakhnov commented on neonvm-daemon/cmd/main.go:
```diff
@@ -26,24 +32,73 @@ func main() {
+	// should be replaced with cgroups milliseconds to be exposed instead of having CPU
+	s.cpuOperationsMutex.Lock()
+	defer s.cpuOperationsMutex.Unlock()
+	activeCPUs, err := s.cpuSystemWideScaler.GetActiveCPUsCount()
```
Will change, but it is not exactly getter

mikhail-sakhnov commented on neonvm-daemon/cmd/main.go:
```diff
@@ -26,24 +32,73 @@ func main() {
+	activeCPUs, err := s.cpuSystemWideScaler.GetActiveCPUsCount()
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
```
great catch, than you, should be logged

mikhail-sakhnov commented on neonvm-runner/cmd/main.go:
```diff
@@ -641,15 +640,12 @@ func newConfig(logger *zap.Logger) *Config {
 	flag.BoolVar(&cfg.skipCgroupManagement, "skip-cgroup-management",
 		cfg.skipCgroupManagement,
 		"Don't try to manage CPU")
-	flag.BoolVar(&cfg.enableDummyCPUServer, "enable-dummy-cpu-server",
```
It will fail with undefined argument. How much is it an issue since we control manifests to deploy?

Omrigan replied:
I guess it doesn't matter, it was set by neonvm-controller, and you've removed it in this PR as well.

And neonvm-runner image is tied to neovm-controller version, so there is no chance older controller would run newer images.

@sharnoff am I correct?

sharnoff replied:
That's correct, yeah. Each neonvm-controller instance only creates runners with the same version.

mikhail-sakhnov commented on pkg/neonvm/controllers/vm_qmp_test.go:
```diff
@@ -1,10 +1,8 @@
-package controllers_test
+package controllers
```
leftover from moving files during development, thanks for catching

mikhail-sakhnov commented on neonvm-daemon/cmd/main.go:
```diff
@@ -26,24 +32,73 @@ func main() {
+	cpuHotPlugController := &cpuscaling.CPUSysFsStateScaler{}
+	srv := cpuServer{
+		cpuOperationsMutex:  &sync.Mutex{},
+		cpuSystemWideScaler: cpuHotPlugController,
```
I agree to rename cpuHotPlugController to something (or actually drop intermediate variable here at all), but I don't want to rename "cpuSystemWideScaler" to "cpuSysfsScaler", may be "cpuScaler"? 

because later there would be also something to ensure cgroup limits and my idea was that in the code it would something like:

```pseudocode
s.cpuScaler.EnsureOnlineCPUs(cpuN) // n is non fractional, may be could be EnsureCPUs
s.cgroupsScaler.SetCGroupLimit(cgroupsN) // cgroupsN is fractional and by design lower or equal to cpuN
```

Omrigan replied:
Yes, `cpuScaler` sounds good, I just wanted `SystemWide` out of the way, to not introduce new concepts.

mikhail-sakhnov commented on neonvm-runner/cmd/main.go:
```diff
@@ -1030,37 +1050,31 @@ func runQEMU(
-		go listenForCPUChanges(ctx, logger, vmSpec.RunnerPort, callbacks, &wg)
+	var callbacks cpuServerCallbacks
+
+	lastValue := &atomic.Uint32{}
```
Discussed on 1o1, we need it until we have fractional cgroups here implemented since controller sends the fractional cgroup limit which is rounded up to the amount of cpu

Omrigan replied:
Yes, right, maybe add a comment here.

mikhail-sakhnov replied:
added

mikhail-sakhnov commented on tests/e2e/tmp-default-cpu-scaling-mode/00-assert.yaml:
```diff
@@ -0,0 +1,13 @@
+kind: VirtualMachine
+metadata:
+  name: example
+# temporary added this check to validate that defaulting cpuScalingMode works as expected
```
>Why is it temporary? To me it seems it should be around as long as the field exists.

I think original idea was to eventually we drop the field, when we will be sure that sys fs scaling works

Omrigan replied:
> not forgetting to delete it

The test would break - impossible to forget. 

I'm just saying explicitly marking it as temporary gives the impression that the test is "more temporary", than the field.

Anyway, up to you.

mikhail-sakhnov replied:
kept it as is

mikhail-sakhnov commented on pkg/neonvm/controllers/vm_controller_cpu_scaling.go:
```diff
@@ -0,0 +1,124 @@
+	log := log.FromContext(ctx)
+	specCPU := vm.Spec.Guest.CPUs.Use
+
+	cgroupUsage, err := getRunnerCgroup(ctx, vm)
```
moved to a separate file and renamed to setRunnerCPULimits/getRunnerCPULimits

mikhail-sakhnov commented on pkg/neonvm/controllers/vm_controller.go:
```diff
@@ -539,21 +548,25 @@ func (r *VMReconciler) doReconcile(ctx context.Context, vm *vmv1.VirtualMachine)
 			}
-
+			var pluggedCPU uint32
+			if vm.Spec.CpuScalingMode != nil && *vm.Spec.CpuScalingMode == vmv1.CpuScalingModeCpuSysfsState {
```
> What's the conclusion of the spec and/or status/annotation discussion, are we leaving this value in spec?

We talked with @sharnoff and decided to leave it in spec and make it immutable if it is not nil

mikhail-sakhnov commented:
@sharnoff @Omrigan I noticed we don't have readiness probe for the vm pod. Wdyt about adding readiness probe to check neonvm-daemon socket availability?

Omrigan commented on pkg/neonvm/controllers/vm_controller_cpu_scaling.go:
```diff
@@ -0,0 +1,121 @@
+// if vm scaling mode is set to CpuSysfsState, the scaling is delegated to neonvm-daemon
+// otherwise the scaling is first done by scaling amount of cores in the VM using QMP and then by updating the cgroup
+func (r *VMReconciler) handleCPUScaling(ctx context.Context, vm *vmv1.VirtualMachine, vmRunner *corev1.Pod) (bool, error) {
+
```
```suggestion
```

Omrigan reviewed (commented):
Didn't do another full pass yet, but a couple of notes:
1. There are couple of unaddressed comments, maybe they slipped from you because they are on now-outdated code. They are visible in "Conversation" tab.
2. Do you plan to merge it with rebase or with squash?

mikhail-sakhnov commented:
@Omrigan 

> Didn't do another full pass yet, but a couple of notes:
> There are couple of unaddressed comments, maybe they slipped from you because they are on now-outdated code. They are visible in "Conversation" tab.

As I mentined, I haven't found any more unaddressed comments.


> Do you plan to merge it with rebase or with squash?

I don't have any specific preference here, what do you suggest?

sharnoff commented on pkg/neonvm/controllers/vm_controller_cpu_scaling.go:
```diff
@@ -0,0 +1,119 @@
+	"github.com/neondatabase/autoscaling/pkg/api"
+)
+
+// handleCPUScaling is extracted from doReconcile and encapsulates the logic to handle CPU scaling.
```
suggestion:

> handleCPUScaling is extracted from doReconcile

IIUC, that's not relevant to what this function is / does, right? that's just the history?

If so, I'd just remove that part, so it reads

> handleCPUScaling encapsulates the logic to handle CPU scaling.

mikhail-sakhnov replied:
rephrased comment

sharnoff commented on pkg/neonvm/controllers/vm_controller_cpu_scaling.go:
```diff
@@ -0,0 +1,119 @@
+	return scaled, nil
+}
+
+// handleCPUScalingQMP handles CPU scaling using QMP, extracted as is from doReconcile
```
Same note here

mikhail-sakhnov replied:
rephrased comment

sharnoff commented on pkg/neonvm/controllers/vm_qmp_test.go:
```diff
@@ -4,7 +4,7 @@ import (
 	. "github.com/onsi/gomega"
 
-	"github.com/neondatabase/autoscaling/pkg/neonvm/controllers"
+	. "github.com/neondatabase/autoscaling/pkg/neonvm/controllers"
```
suggestion: IIUC this change is unrelated? if so, let's remove it and the other changes to this file?

mikhail-sakhnov replied:
reverted that change

sharnoff commented on pkg/neonvm/controllers/vm_controller.go:
```diff
@@ -1474,10 +1382,9 @@ func podSpec(
 						cmd := []string{"runner"}
 						if config.DisableRunnerCgroup {
-							cmd = append(cmd, "-skip-cgroup-management")
 							// cgroup management disabled, but we still need something to provide
 							// the server, so the runner will just provide a dummy implementation.
```
question: Is this comment still valid now that `-enable-dummy-cpu-server` was removed?

mikhail-sakhnov replied:
droped comment

sharnoff commented on neonvm/apis/neonvm/v1/virtualmachine_webhook.go:
```diff
@@ -145,6 +145,11 @@ func (r *VirtualMachine) ValidateUpdate(old runtime.Object) (admission.Warnings,
 
+	// allow to change CPU scaling mode only if it's not set
+	if r.Spec.CpuScalingMode != nil && before.Spec.CpuScalingMode != nil && *r.Spec.CpuScalingMode != *before.Spec.CpuScalingMode {
+		return nil, fmt.Errorf(".spec.cpuScalingMode is immutable")
```
suggestion: not quite "immutable" per se, but can only be set once, right? If so, should this message be different?

mikhail-sakhnov replied:
changed message

sharnoff commented on neonvm/apis/neonvm/v1/virtualmachine_webhook.go:
```diff
@@ -145,6 +145,11 @@ func (r *VirtualMachine) ValidateUpdate(old runtime.Object) (admission.Warnings,
 		}
 	}
 
+	// allow to change CPU scaling mode only if it's not set
+	if r.Spec.CpuScalingMode != nil && before.Spec.CpuScalingMode != nil && *r.Spec.CpuScalingMode != *before.Spec.CpuScalingMode {
```
question: This still allows (non-nil) -> (nil), right? is that intentional?

mikhail-sakhnov replied:
Nope, it is a great catch in the logic flaw, thanks for pointing out

sharnoff commented on neonvm/apis/neonvm/v1/virtualmachine_types.go:
```diff
@@ -150,6 +158,11 @@ type VirtualMachineSpec struct {
 	TargetRevision *RevisionWithTime `json:"targetRevision,omitempty"`
+
+	// Controls how CPU scaling is performed, either hotplug new CPUs with QMP, or enable them in sysfs.
+	// +kubebuilder:validation:Enum=qmpScaling;sysfsScaling
+	// +optional
+	CpuScalingMode *string `json:"cpuScalingMode,omitempty"`
```
suggestion: I would rather use a distinct type here, instead of `string`. IIRC you can then do the `kubebuilder:validation:Enum` on the type definition (see e.g. `RestartPolicy` in this file)

(with that, a small nit: the values should probably be upper camel case)

mikhail-sakhnov replied:
changed it to be type + used kubebuilder:validation:Enum at the type level

sharnoff replied:
nice, looks good!

sharnoff commented on pkg/neonvm/cpuscaling/sysfs.go:
```diff
@@ -0,0 +1,195 @@
+		return err
+	}
+
+	if onlineCount < uint32(targetCount) {
+		for _, cpu := range cpus {
+			if cpu == 0 {
```
suggestion: this is a lot of nested control flow. Maybe worth separating into >1 function?

*(recreated from https://github.com/neondatabase/autoscaling/pull/1111#discussion_r1804084387)*

mikhail-sakhnov replied:
Changed the code in this layer:
- split into two structures, one that reconciles the amount of CPUs and the other one works with sysfs files
- added unit tests for the reconcilation logic
- changed functions to work with ranges instead of slices.

sharnoff commented on pkg/neonvm/cpuscaling/sysfs.go:
```diff
@@ -0,0 +1,195 @@
+}
+
+// parseCPURange parses the CPU range string (e.g., "0-3") and returns a list of CPUs.
+func parseCPURange(cpuRange string) ([]int, error) {
```
suggestion: Should this return (start ID, end ID, error) rather than ([]IDs, error) ?

*(recreated from https://github.com/neondatabase/autoscaling/pull/1111#discussion_r1804084900)*

mikhail-sakhnov replied:
It is not always possible, input might be for example "0-13,15,17-19". The logic of this particular code is that we always enable/disable in order, but in general the CPU state is shared resource and there are no guarantees that it is kept in order.

mikhail-sakhnov replied:
changed

sharnoff commented on pkg/neonvm/controllers/vm_controller_cpu_scaling.go:
```diff
@@ -0,0 +1,119 @@
+}
+
+// handleCPUScalingQMP handles CPU scaling using QMP, extracted as is from doReconcile
+func (r *VMReconciler) handleCPUScalingQMP(ctx context.Context, vm *vmv1.VirtualMachine, vmRunner *corev1.Pod) (bool, error) {
```
suggestion: This function is very visually dense. Could you break it up more by adding newlines between sections?

mikhail-sakhnov replied:
tried to improve readability here.

sharnoff reviewed (commented):
Some comments — looks mostly there!

A couple higher level things I'd note:

* There are some unrelated changes in here that make it harder to review and unnecessarily delay those changes (see e.g. `Makefile` changes, `pkg/neonvm/controllers/vm_controller_unt_test.go`) — if it's not too much effort, it'd be useful to separate those out. Not strictly required to merge though.

* For things like "moving the CPU QMP scaling logic out of `doReconcile`", it would be helpful to be able to view that as a separate commit. This is of course much more involved, so I don't think it's required to merge, but for the future — I often find that keeping refactoring separate from other changes is really helpful for reviewing :sweat_smile:

mikhail-sakhnov commented on pkg/neonvm/cpuscaling/sysfsscaling.go:
```diff
@@ -0,0 +1,196 @@
+		return err
+	}
+
+	if onlineCount < uint32(X) {
```
Is it still applicable keeping in mind changes from the other comment https://github.com/neondatabase/autoscaling/pull/1111#discussion_r1840573281?

sharnoff commented on pkg/neonvm/cpuscaling/cpuscaler.go:
```diff
@@ -0,0 +1,124 @@
+	return nil
+}
+
+// ActiveCPUsCount() returns the count of online CPUs.
+func (c *CPUScaler) ActiveCPUsCount() (int, error) {
```
Just discovered `/sys/devices/system/cpu/online` — would that work here?

mikhail-sakhnov replied:
it coulds be used as well, that's correct. I found that checking each cpu is more straightforward approach then parsing ranges in this file because format is a bit of a mix of ranges and values and could be contrintuitive. For example, value like `0,3,7-10,13-15,17-19` is totally legit.

mikhail-sakhnov replied:
on the other hand we still use range parsing to get list of available CPUs, so might be used just the same here. Let me double-check tomorrow, why I don't use it at the moment, I remember that originally I used this file, but switched to the per-cpu files during testing.

sharnoff commented on pkg/neonvm/cpuscaling/cpuscaler.go:
```diff
@@ -0,0 +1,98 @@
+package cpuscaling
```
question: IIUC this package only works with sysfs scaling, is there a reason for that?

(asking in part because it seems like our sysfs usage could be a lot more efficient, if not for the definition of the `CPUStater` interface)

mikhail-sakhnov replied:
The current package name came from the interface rather than implementation, it feels more or less natural to me to name packages after what they do, not how they do. I can rename the package, if that is not reasonable for you.

>asking in part because it seems like our sysfs usage could be a lot more efficient, if not for the definition of the CPUStater interface

I'm not sure I understand the statement. Could you please clarify? The interface is defined as a separate part that actually works with the filesystem to allow unit testing of the scaling logic.

sharnoff replied:
For some context, the reason I'm pushing on this is that I'm wary about extra complexity — we have a whole setup where we're looping through CPUs, checking their states, and keeping a running counter while we maybe update them, but we already have an easy way of finding just the CPUs we want to update — why not just use that!

sharnoff commented on pkg/neonvm/cpuscaling/sysfs.go:
```diff
@@ -0,0 +1,127 @@
+}
+
+// parseMultipleCPURange parses the multiple CPU range string (e.g., "0-3,5-7") and returns a list of CPUs.
+func (cs *cpuSysfsState) parseMultipleCPURange(cpuRange string) ([]int, error) {
```
```suggestion
func (cs *cpuSysfsState) parseMultipleCPURange(cpuRanges string) ([]int, error) {
```

(because it's not just one range ?)

sharnoff commented on pkg/neonvm/cpuscaling/sysfs.go:
```diff
@@ -0,0 +1,127 @@
+}
+
+// ActiveCPUsCount() returns the count of online CPUs.
+func (c *cpuSysfsState) ActiveCPUsCount() (int, error) {
```
Here, can we just
```go
ranges, err := os.ReadFile("/sys/devices/system/cpu/online")
if err != nil {
    ...
}

cpus, err := c.parseMultipleCPURange(ranges)
if err != nil {
    ...
}

return len(cpus), nil
```

?

mikhail-sakhnov replied:
changed to use intermediate getAllOnlineCPUs which is used in GetState

sharnoff commented on pkg/neonvm/controllers/vm_controller.go:
```diff
@@ -81,8 +78,7 @@ type VMReconciler struct {
 	Config   *ReconcilerConfig
-
-	Metrics ReconcilerMetrics `exhaustruct:"optional"`
+	Metrics  ReconcilerMetrics `exhaustruct:"optional"`
```
nit: unrelated diff?

mikhail-sakhnov replied:
reverted

sharnoff commented on neonvm-daemon/cmd/main.go:
```diff
@@ -26,24 +32,71 @@ func main() {
 }
 
-type cpuServer struct{}
+func (s *cpuServer) handleGetCPUStatus(w http.ResponseWriter) {
```
same comment re: whitespace on `handleGetCPUStatus` and `handleSetCPUStatus` — just a couple blank lines to provide visual distinction to the different parts should suffice

sharnoff commented on pkg/neonvm/cpuscaling/sysfs.go:
```diff
@@ -0,0 +1,127 @@
+	return cs.parseCPURange(string(data))
+}
+
+// parseCPURange parses the CPU range string (e.g., "0-3") and returns a list of CPUs.
```
> returns a list of CPUs.

question: This is no longer true, right? Let's change it?

mikhail-sakhnov replied:
changed

sharnoff commented on pkg/neonvm/controllers/vm_controller.go:
```diff
@@ -165,6 +161,17 @@ func (r *VMReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Re
+	// examine cpuScalingMode and set it to the default value if it is not set
+	if vm.Spec.CpuScalingMode == nil || *vm.Spec.CpuScalingMode == "" {
+		log.Info("Setting default CPU scaling mode", "default", r.Config.DefaultCPUScalingMode)
+		vm.Spec.CpuScalingMode = lo.ToPtr(vmv1.CpuScalingMode(r.Config.DefaultCPUScalingMode))
```
question: Is the `vmv1.CpuScalingMode(...)` cast necessary here?

mikhail-sakhnov replied:
dropped

sharnoff commented on pkg/neonvm/controllers/vm_controller.go:
```diff
@@ -165,6 +161,17 @@ func (r *VMReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Re
 	}
 
+	// examine cpuScalingMode and set it to the default value if it is not set
+	if vm.Spec.CpuScalingMode == nil || *vm.Spec.CpuScalingMode == "" {
```
question: `*vm.Spec.CpuScalingMode == ""` can't happen, right?

mikhail-sakhnov replied:
Yes, a leftover from the field not being valiated

sharnoff commented on pkg/neonvm/controllers/vm_controller.go:
```diff
@@ -553,18 +560,35 @@ func (r *VMReconciler) doReconcile(ctx context.Context, vm *vmv1.VirtualMachine)
+			}
+
+			switch *vm.Spec.CpuScalingMode {
+			case vmv1.CpuScalingModeSysfs:
+				log.Info("CPU usage check based on cgroups", "CpuScalingMode", *vm.Spec.CpuScalingMode)
```
question: What's going on here? Saw "sysfs" + "based on cgroups" and was confused :sweat_smile:

mikhail-sakhnov replied:
yeah, it is a consequence of the fact that all helper functions to call runner used to be called with Cgroups in it, like setRunnerGroups/getRunnerGroups, etc and eventually, it reached to the current state. Changed the log message.

mikhail-sakhnov replied:
In fact, I don't think we really need those logs at all

sharnoff reviewed (commented):
One more round. I think after this it's likely to be good to go.

mikhail-sakhnov commented on pkg/neonvm/controllers/vm_controller_cpu_scaling.go:
```diff
@@ -0,0 +1,127 @@
+			return false, err
+		}
+	} else {
+		log.Info("No need to plug or unplug CPU")
```
dropped

sharnoff commented on pkg/neonvm/cpuscaling/cpuscaler.go:
```diff
@@ -0,0 +1,91 @@
+	if len(online) > targetCount {
+		diff := len(online) - targetCount
+		// offline 'diff' CPUs that are currently online
+		// reverse online slice
```
> reverse online slice

The comment says what this is doing, but not why — could you add that?

sharnoff reviewed (approved):
Looks great! :rocket:

Left just a couple final comments

mikhail-sakhnov commented on pkg/neonvm/cpuscaling/sysfs.go:
```diff
@@ -0,0 +1,102 @@
+	cpuIDs, err := cs.parseMultipleCPURange(string(data))
+	if err != nil {
+		// log value of the file in case we can't parse to help debugging
+		return nil, fmt.Errorf("failed to parse online CPUs `%s`: %w", string(data), err)
```
changed

Omrigan commented on neonvm-runner/cmd/main.go:
```diff
@@ -890,12 +886,32 @@ func buildQEMUCmd(
+
+	switch cfg.cpuScalingMode {
+	case vmv1.CpuScalingModeSysfs:
+		qemuCmd = append(qemuCmd, "-smp", fmt.Sprintf(
+			// Boot with all CPUs plugged, we will online them on-demand
```
```suggestion
		// Boot with all CPUs plugged, we will online them on-demand
		qemuCmd = append(qemuCmd, "-smp", fmt.Sprintf(
```

To match the format below.

Omrigan reviewed (approved):
LGTM, one suggestion.
